### PR TITLE
tee/prototype: Return processing errors for ProcessTXs/Blocks

### PR DESCRIPTION
**Pull request by sebastianst**
Before, these functions just pushed the request into the processors'
queues and returned nil immediately. Now they block until all transactions
or blocks are processed and returns all errors, if any, that occurred
while processing.

### CODE REVIEW

File: tee/prototype/epochproc.go [hunk 1]
```diff
@@ -120,9 +124,12 @@ func (e *Enclave) txRoutine(
 			})
 			log.Debug("txRoutine: TX received, applying...")
 			stagedTxs.cacheTx(tx)
-			err := e.applyEpochTx(txEpoch, tx)
-			if err != nil {
+			if err := e.applyEpochTx(txEpoch, tx); err != nil {
```
Comment (ndzik) [inline tee/prototype/epochproc.go:127]: I remember now, that our `applyEpochTx(...)` makes it way to easy to DOS our system. Unless we require the operator to pre-validate all TXs before pushing them into the Enclave, we shutdown way to strict in that regard.

I propose to have a Enclave.Global `error` type which signals a hard failure. Since we `wrap` our errors, we can easily check with `errors.Is(returnedErr, EnclaveAbortErr)` whether we need to stop the enclave or not. Would also allow to keep logging all errors and remove the workaround present in `applyEpochTx(...)` where we only allow mismatched epoch numbers to occur.

But this would probably be a new issue then.

Comment (sebastianst) [reply tee/prototype/epochproc.go:127]: The whole enclave design is not production ready as it is right now anyways ;) it's just still a demo, after all. I propose to discuss fundamental changes to how we deal with invalid txs in a new issue. This PR should only provide one thing, error/result receipts to the caller of `ProcessTXs`/`ProcessBlocks`.

(Pre-validation has the big disadvantage that we'd check all signatures twice on the same machine, since the op would share computational ressorces with the TEE. In a production setting this could be solved by a layered approach, where one part of the operator is running on a different machine, just receiving and validating txs. But this only becomes relevant in the far future, I guess ;) )

File: client/test/client.go [hunk 2]
```diff
@@ -112,7 +133,90 @@ func (c *Client) Send(recipient common.Address, amount *big.Int) error {
...
+	txs = append(txs, &tx0)
+
+	signAppend := func(tx *tee.Transaction) {
+		c.SignTx(tx)
+		txs = append(txs, tx)
+	}
+
+	// 1. Invalid nonce
+	tx1 := valid
+	tx1.Nonce += 42
```
Comment (ndzik) [inline client/test/client.go]: I have the feeling it might be better to test the edge cases, where the `Nonce` (and everywhere else, where applicable) is exactly `1` more and/or less than expected.

Since off-by-one errors are more common and this would also catch places where an implementer mixed up `{<|>|>=|<=}`.

File: client/test/client.go [hunk 3]
```diff
@@ -112,7 +133,90 @@ func (c *Client) Send(recipient common.Address, amount *big.Int) error {
...
+	signAppend(&tx1)
+
+	// 2. Invalid epoch
+	tx2 := valid
+	tx2.Epoch += 42
```
Comment (ndzik) [inline client/test/client.go]: See other comment for off-by-one errors.

File: eth/wallet.go [hunk 4]
```diff
@@ -21,3 +24,7 @@ func NewHdWallet(rng io.Reader) *hdwallet.Wallet {
 	}
 	return hdw
 }
+
+func NewRandomAddress(rng *rand.Rand) common.Address {
```
Comment (ndzik) [inline eth/wallet.go]: This function is also already implemented in `tee/test/transaction.go`. Would be nice to remove it from there if possible.

Would need to touch the `txCache` test then IIRC.

**Review by ndzik — CHANGES_REQUESTED** [review]

File: client/test/client.go (hunk 2 above)
Comment (sebastianst) [reply client/test/client.go]: good point, will do that

File: eth/wallet.go (hunk 4 above)
Comment (sebastianst) [reply eth/wallet.go]: Ah forgot about that one, will remove my implementation then.

File: client/test/client.go (hunk 2 above)
Comment (sebastianst) [reply client/test/client.go]: This inspired me to also check for negative amounts and we didn't check that yet :joy: fixed now.

**Review by ndzik — APPROVED** [review]
